packages/circuits/src/initial.circom

```diff
@@ -0,0 +1,89 @@
+pragma circom 2.0.0;
+
+include "./merkle.circom";
+include "../node_modules/circomlib/circuits/comparators.circom";
+
+template CheckReplaceArray(n) {
+    signal input index;
+    signal input newValue;
+    signal input array[n];
+    signal output newArray[n];
+    signal output replacedValue;
+
+    component checkIndex[n];
+    signal items[n];
+    for (var i = 0; i < n; i++) {
+        checkIndex[i] = IsEqual();
+        checkIndex[i].in[0] <== i;
+        checkIndex[i].in[1] <== index;
+        newArray[i] <== (newValue - array[i]) * checkIndex[i].out + array[i];
+        // newArray[i] <== (checkIndex[i].out * newValue) + ((1 - checkIndex[i].out) * array[i]);
+        items[i] <== checkIndex[i].out * array[i];
+    }
+
+    // get the replaced value
+    var temp = 0;
+    for (var i = 0; i < n; i++) {
+        temp += items[i];
+    }
+    replacedValue <== temp;
```

Comment:
Is this safe? My understanding of circom is that you could set temp to anything you want here (i.e. it is not bound to `array[index]`).

Can't we simply include `replacedValue <== array[index]` here?

Reply:
Ah! Maybe it's secure because `var`is under the hood an array of signals that get constrained to their next value. In any case, the question about the simpler form still holds.

Reply:
We can't use `replacedValue <== array[index]` because `index` is an unknown value during compilation, where in Circom any reading/writing to an array couldn't use an unknown value, otherwise there will be non-quadratic constraint error. I do feel that my approach of looping through the array and summing to `temp` is pretty dumb, but I couldn't think of a better way, so if there's other ideas or inspirations that would be appreciated too.

Reply:
Right, that seems like a pretty fundamental limitation. Circomlib has [`Mux4`](https://github.com/iden3/circomlib/blob/master/circuits/mux4.circom) that allows you to index into an array of size 16, but that wouldn't work for the deck.

One question is whether we need `replacedValue` at all. You use it here: 

```
updateDeckArray[i+1].newValue <== updateDeckArray[i].replacedValue;
```

But I think we could skip this if we supplied the list of drawn cards as a signal. We'd need an additional constraint to check that the card matches the index, but we'd avoid O(deck_size) extra constraints for getting the replaced value.

Reply:
If that's the case, we will need to pass in both the drawn cards and tail cards as signals, as we need both to check that the decks have been updated correctly. Could again do a performance comparison.